### DISCUSSION TITLE
avcodec/decode.c/avcodec_open2() fails due to missing AVCodecContext extradata - part 2

### DISCUSSION THREAD

**Discussion by KORuL** [baresip/baresip#2245] [General]
In continuation of this topic
https://github.com/baresip/baresip/issues/985

https://github.com/baresip/baresip/issues/985#issuecomment-636304160

Good evening! Tell me how to do it.
I built ffmpeg with `--enable-android-media-codec` flag
In **baresip.c** `JNI_OnLoad` added `av_jni_set_java_vm(vm, NULL);`
added to config
```
#avcodec
avcodec_h264dec h264_mediacodec
avcodec_hwaccel mediacodec
```
and have the following problem, decoder initialization problem. Tell me how to be

**Comment by alfredh** [comment]
this is a missing feature

You have to extract SPS and PPS, convert it to AVCC format and then pass it to extradata.
It has been added to the roadmap:

https://github.com/baresip/baresip/wiki/Roadmap

A Pull Request is welcome

**Comment by KORuL** [comment]
I understand what is needed, I do not understand how to do it

**Reply by alfredh** [reply]
You can try this function:

https://github.com/baresip/rem/blob/main/include/rem_avc.h

**Reply by KORuL** [reply]
This is a conversion to avcc format, but where and how to get sps pps

**Comment by frogblueDFi** [comment]
Hi!

I am very happy with our iOS SIP app based on baresip used as library.
Using VideoToolbox hwaccel for h264/h265 encoding is functional. 

We are now close to an Android version.
What is the state of the art for supporting Androids MediaCodec hwaccel?

I have implemented it to the point where ffmpeg fails in avcodec.c avcodec_open2 function.

Debug output up to here:

```
avcodec: module_init
avcodec: using H.264 encoder 'h264_mediacodec' -- H.264 Android MediaCodec encoder
avcodec: using H.264 decoder 'h264_mediacodec' -- H.264 Android MediaCodec decoder
avcodec: using H.265 encoder 'hevc_mediacodec' -- H.265 Android MediaCodec encoder
avcodec: using H.265 decoder 'hevc_mediacodec' -- H.265 Android MediaCodec decoder
avcodec: check using hwaccel
avcodec: enable hwaccel using 'mediacodec'
avcodec: decode: using hardware pixel format 'mediacodec'
avcodec: FFMPEG (Disable createPersistentInputSurface)
avcodec: h264 decoder activated
avcodec: decode: hardware accel enabled (mediacodec)
avcodec: FFMPEG (The input looks like it is Annex B already)
avcodec: decode: could not open codec: (-1)
avcodec: H264: could not init decoder
```

Sorry, I have no idea what "extract SPS and PPS, convert it to AVCC format" means.
Found the corresponding comment in baresip avcodec/decode.c.
Something missing in baresip implementation?
Did anyone get Android MediaCodec hwaccel to work?

Thank you for any hints!
Dietmar
